test_math_util_length_domain fails on CPython 3.13

```
> python3.13 -m pytest spherical_geometry/tests/test_basic.py -k test_math_util_length_domain
spherical_geometry/tests/test_basic.py F                                 [100%]

=================================== FAILURES ===================================
_________________________ test_math_util_length_domain _________________________
ValueError: Out of domain for acos

The above exception was the direct cause of the following exception:

    def test_math_util_length_domain():
        with pytest.raises(ValueError):
>           great_circle_arc.length([[np.nan, 0, 0]], [[0, 0, np.inf]])

spherical_geometry/tests/test_basic.py:533:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

A = [[nan, 0, 0]], B = [[0, 0, inf]]

    def length(A, B):
        r"""
        Returns the angular distance between two points (in vector space)
        on the unit sphere.

        Parameters
        ----------
        A, B : (*x*, *y*, *z*) triples or Nx3 arrays of triples
           The endpoints of the great circle arc, in vector space.

        Returns
        -------
        length : scalar or array of scalars
            The angular length of the great circle arc in radians.

        Notes
        -----
        The length is computed using the following:

        .. math::

           \Delta = \arccos(A \cdot B)
        """
        if HAS_C_UFUNCS:
>           result = math_util.length(A, B)
E           SystemError: <class 'RuntimeWarning'> returned a result with an exception set

spherical_geometry/great_circle_arc.py:210: SystemError
=========================== short test summary info ============================
FAILED spherical_geometry/tests/test_basic.py::test_math_util_length_domain - SystemError: <class 'RuntimeWarning'> returned a result with an exception set
======================= 1 failed, 31 deselected in 0.05s =======================
```

All tests pass on 3.12

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -4,6 +4,14 @@
 Release Notes
 =============
 
+1.3.3 (30-January-2025)
+=======================
+
+- Fixed a bug in the C implementation of ``dot_qd`` that may result in
+  "invalid value encountered in length" runtime warning when input vectors
+  contain NaN values. [#288]
+
+
 1.3.2 (12-June-2024)
 ====================
 
@@ -15,6 +23,7 @@ Release Notes
   complication. [#265]
 
 - Build wheel with Numpy 2.0 release candidate ahead of Numpy 2.0 release [#279]
+
 
 1.3.1 (5-November-2023)
 =======================

--- a/spherical_geometry/tests/test_basic.py
+++ b/spherical_geometry/tests/test_basic.py
@@ -548,9 +548,8 @@ def test_math_util_length_domain():
         [0, 0, 0],
     ]
 
-    v = great_circle_arc.length(a, b)
-
-    assert np.all(np.logical_not(np.isfinite(v)))
+    with pytest.raises(ValueError):
+        great_circle_arc.length(a, b)
 
 
 def test_math_util_angle_nearly_coplanar_vec():

--- a/spherical_geometry/tests/test_basic.py
+++ b/spherical_geometry/tests/test_basic.py
@@ -529,8 +529,28 @@ def test_math_util_angle_domain():
 
 
 def test_math_util_length_domain():
-    with pytest.raises(ValueError):
-        great_circle_arc.length([[np.nan, 0, 0]], [[0, 0, np.inf]])
+    a = [
+        [np.nan, 0, 0],
+        [np.nan, 0, 0],
+        [np.nan, np.nan, np.nan],
+        [0, 0, 0],
+        [0, 0, 0],
+        [0, 0, np.nan],
+        [0, 0, 0],
+    ]
+    b = [
+        [0, 0, np.inf],
+        [0, 0, np.inf],
+        [0, 0, 0],
+        [np.inf, np.inf, np.inf],
+        [0, 0, np.inf],
+        [0, 0, 0],
+        [0, 0, 0],
+    ]
+
+    v = great_circle_arc.length(a, b)
+
+    assert np.all(np.logical_not(np.isfinite(v)))
 
 
 def test_math_util_angle_nearly_coplanar_vec():

--- a/src/math_util.c
+++ b/src/math_util.c
@@ -176,6 +176,10 @@ dot_qd(const qd *A, const qd *B, qd *C) {
     double tmp[4][4];
 
     for (i = 0; i < 3; ++i) {
+        if (ISNAN_QD(A[i]) || ISNAN_QD(B[i])) {
+            c_qd_copy_d(NPY_NAN, C->x);
+            return;
+        }
         c_qd_mul(A[i].x, B[i].x, tmp[i]);
     }
 

--- a/src/math_util.c
+++ b/src/math_util.c
@@ -146,11 +146,6 @@ normalize_qd(const qd *A, qd *B) {
     double l[4];
 
     for (i = 0; i < 3; ++i) {
-        if (ISNAN_QD(A[i])) {
-            c_qd_copy_d(NPY_NAN, B->x);
-            return 1;
-        }
-
         c_qd_sqr(A[i].x, T[i]);
     }
 
@@ -181,15 +176,12 @@ dot_qd(const qd *A, const qd *B, qd *C) {
     double tmp[4][4];
 
     for (i = 0; i < 3; ++i) {
-        if (ISNAN_QD(A[i]) || ISNAN_QD(B[i])) {
-            c_qd_copy_d(NPY_NAN, C->x);
-            return;
-        }
         c_qd_mul(A[i].x, B[i].x, tmp[i]);
     }
 
     c_qd_add(tmp[0], tmp[1], tmp[3]);
     c_qd_add(tmp[3], tmp[2], C->x);
+    PyUFunc_clearfperr();
 }
 
 /*
@@ -197,7 +189,6 @@ dot_qd(const qd *A, const qd *B, qd *C) {
 */
 static NPY_INLINE int
 normalized_dot_qd(const qd *A, const qd *B, qd *dot_val) {
-    int flag;
     qd aa, bb, ab;
     double aabb[4];
     double norm[4];
@@ -215,7 +206,7 @@ normalized_dot_qd(const qd *A, const qd *B, qd *dot_val) {
         return 1;
     }
 
-    flag = c_qd_sqrt(aabb, norm);
+    c_qd_sqrt(aabb, norm);
 
     if (norm[0] == 0.0) {
         /* return non-normalized value: */
@@ -661,7 +652,7 @@ DOUBLE_length(char **args, const intp *dimensions, const intp *steps, void *NPY_
 #else
             *((double *)op) = strtod("NaN", NULL);
 #endif
-            continue;
+            return;
         }
 
         *((double *)op) = s.x[0];

--- a/src/math_util.c
+++ b/src/math_util.c
@@ -646,15 +646,15 @@ DOUBLE_length(char **args, const intp *dimensions, const intp *steps, void *NPY_
         load_point_qd(ip1, is1, A);
         load_point_qd(ip2, is2, B);
 
-        if (normalize_qd(A, A) || normalize_qd(B, B) || length_qd(A, B, &s)) {
-#if defined(NAN)
-            *((double *)op) = NAN;
-#else
-            *((double *)op) = strtod("NaN", NULL);
-#endif
+        if (normalize_qd(A, A)) {
             return;
         }
-
+        if (normalize_qd(B, B)) {
+            return;
+        }
+        if (length_qd(A, B, &s)) {
+            return;
+        }
         *((double *)op) = s.x[0];
     END_OUTER_LOOP
 

--- a/src/math_util.c
+++ b/src/math_util.c
@@ -181,6 +181,15 @@ dot_qd(const qd *A, const qd *B, qd *C) {
 
     c_qd_add(tmp[0], tmp[1], tmp[3]);
     c_qd_add(tmp[3], tmp[2], C->x);
+
+    // In Python 3.13 it seems that the code above sets a floating point error
+    // flag (when input vectors contain nan/inf values) and this raises a
+    // warning/error "RuntimeWarning: invalid value encountered in length"
+    // and which results in a SystemError once
+    // PyErr_SetString(PyExc_ValueError, "Out of domain for acos") is called
+    // in length_qd. This clears FP error flags before raising the above
+    // exception.
+    // Also See https://github.com/spacetelescope/spherical_geometry/pull/288
     PyUFunc_clearfperr();
 }
 

--- a/src/math_util.c
+++ b/src/math_util.c
@@ -146,6 +146,11 @@ normalize_qd(const qd *A, qd *B) {
     double l[4];
 
     for (i = 0; i < 3; ++i) {
+        if (ISNAN_QD(A[i])) {
+            c_qd_copy_d(NPY_NAN, B->x);
+            return 1;
+        }
+
         c_qd_sqr(A[i].x, T[i]);
     }
 
@@ -650,10 +655,14 @@ DOUBLE_length(char **args, const intp *dimensions, const intp *steps, void *NPY_
         load_point_qd(ip1, is1, A);
         load_point_qd(ip2, is2, B);
 
-        if (normalize_qd(A, A)) return;
-        if (normalize_qd(B, B)) return;
-
-        if (length_qd(A, B, &s)) return;
+        if (normalize_qd(A, A) || normalize_qd(B, B) || length_qd(A, B, &s)) {
+#if defined(NAN)
+            *((double *)op) = NAN;
+#else
+            *((double *)op) = strtod("NaN", NULL);
+#endif
+            continue;
+        }
 
         *((double *)op) = s.x[0];
     END_OUTER_LOOP

